src/sbvr-api/sbvr-utils.coffee

```diff
@@ -353,18 +358,59 @@ getHooks = do ->
 	_getHooks = (request) ->
 		if request.resourceName?
 			resourceName = resolveSynonym(request)
-		getMethodHooks(request.method, request.vocabulary, resourceName)
+		instantiateHooks(
+			getMethodHooks(
+				request.method
+				request.vocabulary
+				resourceName
+			)
+		)
 	_getHooks.clear = -> getMethodHooks.clear()
 	return _getHooks
 
-runHook = (hookName, args) ->
+class Hook
+	constructor: (hookFn) ->
+		@hookFn = hookFn
+		@executed = false
+
+	run: (args...) ->
+		Promise.try =>
+			@hookFn(args...)
+		.finally =>
+			@executed = true
+
+class SideEffectHook extends Hook
+	constructor: (hookFn) ->
+		@rollbackFns = []
+		@rolledBack = false
+		super(hookFn)
+
+	registerRollback: (fn) ->
+		runFn = Promise.method(fn)
+		if @rolledBack
+			runFn()
+		else
+			@rollbackFns.push(runFn)
+
+	rollback: ->
+		# set rolledBack to true straight away, so that if any rollback action
+		# is registered after the rollback call, we will immediately execute it
+		@rolledBack = true
+		Promise.each(@rollbackFns, (fn) -> fn())
```

Comment:
I think this should be a `Promise.map`, otherwise if one of the rollbacks fails then none of the ones after it will be run